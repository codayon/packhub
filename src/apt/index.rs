use std::io::Write;

use anyhow::{bail, Result};
use askama::Template;
use chrono::Utc;
use libflate::gzip::{EncodeOptions, Encoder, HeaderBuilder};
use md5::Md5;
use sha1::Sha1;
use sha2::{Sha256, Sha512};

use crate::{apt::deb::DebAnalyzer, package::Package, utils::hashsum};

pub struct AptIndices<'a> {
    packages: &'a [Package],
}

#[derive(Template)]
#[template(path = "Release")]
struct ReleaseIndex<'a> {
    origin: &'a str,
    label: &'a str,
    date: String,
    files: Vec<Files>,
}

#[derive(Template)]
#[template(path = "Packages")]
struct PackageIndex {
    packages: Vec<DebianPackage>,
}

struct DebianPackage {
    control: String,
    md5: String,
    sha1: String,
    sha256: String,
    sha512: String,
    size: usize,
    filename: String,
}

struct Files {
    md5: String,
    sha1: String,
    sha256: String,
    sha512: String,
    size: usize,
    path: String,
}

impl<'a> AptIndices<'a> {
    pub fn new(packages: &'a [Package]) -> Result<AptIndices<'a>> {
        Ok(AptIndices { packages })
    }

    pub fn get_package_index(&self) -> String {
        let packages = self
            .packages
            .iter()
            .map(|package| get_package_metadata(package))
            .filter_map(Result::ok)
            .collect::<Vec<_>>();
        let index = PackageIndex { packages };
        index.render().unwrap().trim().to_owned()
    }

    pub fn get_release_index(&self) -> String {
        let date = Utc::now().to_rfc2822();

        let packages = self.get_package_index();
        let packages = packages.as_bytes();

        let name = ". stable"; //format!("{} stable", self.deb.get_package());

        let packages_gz = gzip_compression(packages);

        let files = vec![
            Files {
                sha256: hashsum::<Sha256>(packages),
                size: packages.len(),
                path: "main/binary-amd64/Packages".to_string(),
                md5: hashsum::<Md5>(packages),
                sha1: hashsum::<Sha1>(packages),
                sha512: hashsum::<Sha512>(packages),
            },
            Files {
                sha256: hashsum::<Sha256>(&packages_gz),
                size: packages_gz.len(),
                path: "main/binary-amd64/Packages.gz".to_string(),
                md5: hashsum::<Md5>(&packages_gz),
                sha1: hashsum::<Sha1>(&packages_gz),
                sha512: hashsum::<Sha512>(&packages_gz),
            },
        ];

        let index = ReleaseIndex {
            date,
            files,
            origin: name,
            label: name,
        };

        index.render().unwrap()
    }
}

pub fn gzip_compression(data: &[u8]) -> Vec<u8> {
    let header = HeaderBuilder::new().modification_time(0).finish();
    let options = EncodeOptions::new().header(header);
    let mut encoder = Encoder::with_options(Vec::new(), options).unwrap();
    encoder.write_all(data).unwrap();

    let gzip = encoder.finish();
    let gzip = gzip.into_result().unwrap();

    gzip
}

fn get_package_metadata(package: &Package) -> Result<DebianPackage> {
    let Some(data) = package.data() else {
        bail!("Package data is not available");
    };
    let deb = DebAnalyzer::new(&data);
    let control = deb.get_control_data().trim_end().to_owned();
    let filename = format!("pool/stable/{}/{}", package.version(), package.file_name());

    let size = data.len();
    let md5 = hashsum::<Md5>(&data);
    let sha1 = hashsum::<Sha1>(&data);
    let sha256 = hashsum::<Sha256>(&data);
    let sha512 = hashsum::<Sha512>(&data);

    Ok(DebianPackage {
        control,
        md5,
        sha1,
        sha256,
        sha512,
        size,
        filename,
    })
}

#[cfg(test)]
mod tests {
    use std::fs;

    use chrono::DateTime;
    use insta::assert_snapshot;

    use super::*;

    #[test]
    fn test_apt_indices() {
        let package = Package::detect_package("OpenBangla-Keyboard_2.0.0-ubuntu20.04.deb", "2.0.0".to_owned(), "https://github.com/OpenBangla/OpenBangla-Keyboard/releases/download/2.0.0/OpenBangla-Keyboard_2.0.0-ubuntu20.04.deb".to_owned(), DateTime::UNIX_EPOCH).unwrap();
        let data = fs::read("data/OpenBangla-Keyboard_2.0.0-ubuntu20.04.deb").unwrap();
        package.set_data(data);

        let packages = vec![package];

        let indices = AptIndices::new(&packages).unwrap();

        // Packages
        let packages = indices.get_package_index();
        assert_snapshot!(packages);

        // Release
        let release = indices.get_release_index();
        insta::with_settings!({filters => vec![
            // Date is a changing value, so replace it with a hardcoded value.
            (r"Date: .+", "Date: [DATE]"),
        ]}, {
            assert_snapshot!(release);
        });
    }

    #[test]
    fn test_multiple_packages() {
        let package1 = Package::detect_package("fcitx-openbangla_3.0.0.deb", "3.0.0".to_owned(), "https://github.com/mominul/pack-exp2/releases/download/3.0.0/fcitx-openbangla_3.0.0.deb".to_owned(), DateTime::UNIX_EPOCH).unwrap();
        let data = fs::read("data/fcitx-openbangla_3.0.0.deb").unwrap();
        package1.set_data(data);

        let package2 = Package::detect_package("ibus-openbangla_3.0.0.deb", "3.0.0".to_owned(), "https://github.com/mominul/pack-exp2/releases/download/3.0.0/ibus-openbangla_3.0.0.deb".to_owned(), DateTime::UNIX_EPOCH).unwrap();
        let data = fs::read("data/ibus-openbangla_3.0.0.deb").unwrap();
        package2.set_data(data);

        let packages = vec![package1, package2];

        let indices = AptIndices::new(&packages).unwrap();

        // Packages
        let packages = indices.get_package_index();
        assert_snapshot!(packages);
        assert_eq!(packages.as_bytes().len(), 2729);
        let packages_gz = gzip_compression(packages.as_bytes());
        assert_eq!(packages_gz.len(), 1105);

        // Release
        let release = indices.get_release_index();
        insta::with_settings!({filters => vec![
            // Date is a changing value, so replace it with a hardcoded value.
            (r"Date: .+", "Date: [DATE]"),
        ]}, {
            assert_snapshot!(release);
        });
    }
}
