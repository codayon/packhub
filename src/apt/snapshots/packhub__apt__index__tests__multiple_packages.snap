---
source: src/apt/index.rs
expression: packages
---
Architecture: amd64
Depends: fcitx5 (>= 5.0.5), libc6 (>= 2.29), libfcitx5config6 (>= 5.0.5), libfcitx5core7 (>= 5.0.5), libfcitx5utils2 (>= 5.0.5), libgcc-s1 (>= 4.2), libqt5core5a (>= 5.15.1), libqt5gui5 (>= 5.0.2) | libqt5gui5-gles (>= 5.0.2), libqt5network5 (>= 5.0.2), libqt5widgets5 (>= 5.0.2), libstdc++6 (>= 9), libzstd1 (>= 1.4.0)
Description: OpenSource Bengali input method
 OpenBangla Keyboard is an OpenSource, Unicode compliant Bengali Input Method for GNU/Linux systems. It&apos;s a full fledged Bengali input method with typing automation tools, includes many famous typing methods such as Avro Phonetic, Probhat, Munir Optima, National (Jatiya) etc.
 .
 Most of the features of Avro Keyboard are present in OpenBangla Keyboard. So Avro Keyboard users will feel at home with OpenBangla Keyboard in Linux.
Homepage: https://openbangla.github.io/
Maintainer: OpenBangla Team <openbanglateam@gmail.com>
Package: fcitx-openbangla
Priority: optional
Section: utils
Version: 3.0.0
Installed-Size: 12763
MD5sum: 45ae5eb935c9b9b5fc4df927217a55e3
SHA1: bff2bedc76feb324ea2db323eaa568f2234b92a5
SHA256: c84f27813eebf33fc2b90d3fc330cea5167d0fa08d0a674ac863d207916a33dd
SHA512: 555d2a214ad5d5bd196df952b84c071ecf46502e04af5a715c0b0c65677becc87b08de91e8385d1ab39a80d2aa51dd6ce66a64e0b806a8c16c4a52e67c431b06
Size: 5484440
Filename: pool/stable/2.0.0/OpenBangla-Keyboard_3.0.0-fcitx.deb

Architecture: amd64
Depends: ibus (>= 1.5.1), libc6 (>= 2.29), libgcc-s1 (>= 4.2), libglib2.0-0 (>= 2.12.0), libibus-1.0-5 (>= 1.5.1), libqt5core5a (>= 5.12.2), libqt5gui5 (>= 5.0.2) | libqt5gui5-gles (>= 5.0.2), libqt5network5 (>= 5.0.2), libqt5widgets5 (>= 5.0.2), libstdc++6 (>= 5.2), libzstd1 (>= 1.3.2)
Description: OpenSource Bengali input method
 OpenBangla Keyboard is an OpenSource, Unicode compliant Bengali Input Method for GNU/Linux systems. It&apos;s a full fledged Bengali input method with typing automation tools, includes many famous typing methods such as Avro Phonetic, Probhat, Munir Optima, National (Jatiya) etc.
 .
 Most of the features of Avro Keyboard are present in OpenBangla Keyboard. So Avro Keyboard users will feel at home with OpenBangla Keyboard in Linux.
 .
Homepage: https://openbangla.github.io/
Maintainer: OpenBangla Team <openbanglateam@gmail.com>
Package: ibus-openbangla
Priority: optional
Section: utils
Version: 3.0.0
Installed-Size: 12824
MD5sum: 50b535cfbedaa8a64272db3e85082ec6
SHA1: 769405dd1e9da69217f97cbe14d1632201f74c5b
SHA256: 2649db66a21a9ce3d915d856781b83b37dcd069e8834b047ddb2d559e7e54277
SHA512: 9360281f13177c7c4e0cb4cff52cbf0a0d3491febebe8b8cbd6e95ce4dc24f25ca6a0398a41905638d1ab0739667b1737412ba497c1c4cd25d60015b0f481c1f
Size: 5525884
Filename: pool/stable/2.0.0/OpenBangla-Keyboard_3.0.0-ibus.deb


